torch/testing/_internal/common_utils.py

```diff
@@ -801,204 +801,304 @@ def is_integral(self, dtype):
+
+        # Doubles uint width when upcasting
+        if high_meta.kind != 'uint' and low_meta.kind == 'uint':
+            width = min(64, max(low_meta.width * 2, high_meta.width))
+
+        # Accounts for complex types being represented as double width
+        if high_meta.kind == 'complex':
+            width *= 2
+
+        # Accounts for bfloat16 x float16
+        if high_meta.rank == 7 and low_meta.rank == 6:
+            width = 32
+
+        # Acquires common dtype
+        common_dtype = getattr(torch, high_meta.kind + str(width))
+
+        # TODO: update this when isclose is implemented for CPU/CUDA bfloat16
+        if common_dtype is torch.bfloat16:
+            common_dtype = torch.float32
+
+        # non-CUDA (CPU, for example) float16 -> float32
+        # TODO: update this when isclose is implemented for CPU float16
+        if (common_dtype is torch.float16 and
+            (a.device != b.device or a.device.type != 'cuda' or
+             b.device.type != 'cuda')):
+            common_dtype = torch.float32
+
+        return common_dtype
+
+    # Checks if two dense tensors are equal(-ish), returning true when they are.
+    # If exact_dtype is true both tensors must have the same dtype.
+    # If exact_device is true both tensors must be on the same device.
+    # Note: tensors on different devices are moved to the CPU to be compared when
+    # exact_device is False.
+    # TODO: default exact_device to True.
+    def compareTensors(self, a, b, *, rtol, atol, equal_nan,
+                       exact_dtype=False, exact_device=False):
+        self.assertTrue(isinstance(a, torch.Tensor) and isinstance(b, torch.Tensor))
+
+        # Validates tensors are on the same device
+        if exact_device:
+            if a.device != b.device:
+                raise Exception("Attempted to compare equality of tensors on"
+                                " different devices. Got device {0} and"
+                                " {1}.".format(a.device, b.device))
+
+        # Compares tensors of different devices on the CPU
+        if a.device != b.device:
+            a = a.cpu()
+            b = b.cpu()
+
+        # Checks size matches
+        if a.size() != b.size():
+            return False
+
+        # Checks dtype (if exact_dtype)
+        if exact_dtype and a.dtype is not b.dtype:
+            return False
+
+        # Acquires atol and rtol
+        # Note: atol can be overriden by a class's precision value. This
+        # legacy behavior allows generated tests to customize their precision.
+        rtol = rtol if rtol is not None else max(self.dtype_precisions.get(a.dtype, (0, 0))[0],
+                                                 self.dtype_precisions.get(b.dtype, (0, 0))[0])
+        atol = atol if atol is not None else max(self.dtype_precisions.get(a.dtype, (0, 0))[1],
+                                                 self.dtype_precisions.get(b.dtype, (0, 0))[1])
+        atol = max(atol, self.precision)
```

Comment:
I'm not seeing handling for integral types. In the previous code there was an `if self.is_integral(dtype): return (0, 0)`

Reply:
ah I see, just the default when nothing found in the `get` call...